File: src/graph/road_reparation.py
n, m = map(int, input().split())
edges = []
cluster = [i for i in range(n + 1)]
size = [1 for i in range(n + 1)]


def same(u, v):
    # print("same", u, v, "->", find(u) == find(v))
    return find(u) == find(v)


def find(x):
    # if cluster[x] == x:
    #     return x
    # cluster[x] = find(cluster[x])
    while x != cluster[x]:
        x = cluster[x]
    return x


def unite(u, v):
    # print("unite", u, v)
    u, v = find(u), find(v)
    if size[u] > size[v]:
        cluster[v] = cluster[u]
        size[u] += size[v]
        size[v] = 0
    else:
        cluster[u] = cluster[v]
        size[v] += size[u]
        size[u] = 0
    # print(cluster[1:])


for i in range(m):
    edges.append(tuple(map(int, input().split())))

edges.sort(key=lambda v: v[2])

ans = 0
for u, v, wt in edges:
    if not same(u, v):
        ans += wt
        unite(u, v)

# print(cluster[1:])
# print(size[1:])
if size[find(1)] == n:
    print(ans)
else:
    print("IMPOSSIBLE")

File: src/graph/road_reparation.rs
fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.split_whitespace();
    let n: usize = iter.next().unwrap().parse().unwrap();
    let m: usize = iter.next().unwrap().parse().unwrap();

    let mut edges = Vec::new();
    let mut cluster = (0..=n).collect::<Vec<usize>>();
    let mut size = vec![1; n + 1];

    for _ in 0..m {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let u: usize = iter.next().unwrap().parse().unwrap();
        let v: usize = iter.next().unwrap().parse().unwrap();
        let wt: i64 = iter.next().unwrap().parse().unwrap();
        edges.push((u, v, wt));
    }

    edges.sort_by_key(|&(_, _, wt)| wt);

    let mut ans = 0;
    for &(u, v, wt) in &edges {
        if !same(u, v, &mut cluster) {
            ans += wt;
            unite(u, v, &mut cluster, &mut size);
        }
    }

    if size[find(1, &mut cluster)] == n {
        println!("{}", ans);
    } else {
        println!("IMPOSSIBLE");
    }
}

fn same(u: usize, v: usize, cluster: &mut [usize]) -> bool {
    find(u, cluster) == find(v, cluster)
}

fn find(x: usize, cluster: &mut [usize]) -> usize {
    if cluster[x] == x {
        return x;
    }
    cluster[x] = find(cluster[x], cluster);
    cluster[x]
}

fn unite(u: usize, v: usize, cluster: &mut [usize], size: &mut [usize]) {
    let root_u = find(u, cluster);
    let root_v = find(v, cluster);
    if size[root_u] > size[root_v] {
        cluster[root_v] = root_u;
        size[root_u] += size[root_v];
        size[root_v] = 0;
    } else {
        cluster[root_u] = root_v;
        size[root_v] += size[root_u];
        size[root_u] = 0;
    }
}
